### ISSUE TITLE
[MNG-5087] Maven 3 dependency resolution fails until maven-metadata-local.xml files (created by maven-invoker-plugin) are deleted

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7097]
**[Chas Emerick](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cemerick)** opened **[MNG-5087](https://issues.apache.org/jira/browse/MNG-5087?redirect=false)** and commented

In one of my Maven projects, dependency resolution will succeed once, then fail for later build attempts:

```
[WARNING] The POM for commons-logging:commons-logging:jar:1.1.1 is missing, no dependency information available
[WARNING] The POM for commons-httpclient:commons-httpclient:jar:3.1 is missing, no dependency information available
[WARNING] The POM for javax.mail:mail:jar:1.4.4 is missing, no dependency information available
```

...and so on, until I delete the `maven-metadata-local.xml` files corresponding to the failing artifacts (e.g. `~/.m2/repository/commons-logging/commons-logging/maven-metadata-local.xml`), which appear to be created by maven-invoker-plugin:install. After those files are deleted, the next `mvn` invocation proceeds properly; the metadata files are restored by that invocation (presumably as part of the process of checking my upstream repositories/mirrors for updated artifacts), and I am again presented with the above errors until I again delete the metadata files.

This is repeatable, even after starting with a completely fresh local repository. Note that Maven 2.2.1 does **NOT** exhibit this problem.

FYI, I'm not using an integration-testing-only local repo http://maven.apache.org/plugins/maven-invoker-plugin/install-mojo.html#localRepositoryPath|as described here, simply because doing so causes the re-downloading of all transitive dependencies (http://maven.apache.org/plugins/maven-invoker-plugin/examples/fast-use.html|unless you want to maintain an integration-specific settings.xml file!!!). I've used the invoker plugin with a variety of other projects in this way with good results (http://github.com/clojure/tools.nrepl|example) -- certainly never encountering a borked local repository in the process like this.

Here's an affected project: https://github.com/cemerick/rummage/tree/1.3.0-compat|the 1.3.0-compat branch of rummage. To reproduce, just clone that repo, checkout `1.3.0-compat`, and:

```
> mvn clean test
1. no error -- can run this and other builds that don't involve maven-invoker-plugin all day w/o problems
> mvn clean integration-test
1. FAIL: "Could not resolve dependencies", with warnings as noted above
> mvn clean test
1. FAIL: "Could not resolve dependencies", with warnings as noted above
```

Once the local repository is broken (by the generation of the `maven-metadata-local.xml` files, AFAICT), no builds will get past the dependency resolution stage.

Running mvn -X reveals lines like this for each artifact that is later apparently not found:

```
[DEBUG] Verifying availability of /Users/chas/.m2/repository/javax/mail/mail/1.4.4/mail-1.4.4.jar from []
```

Of course, `/Users/chas/.m2/repository/javax/mail/mail/1.4.4/mail-1.4.4.jar` et al. does exist, as does `/Users/chas/.m2/repository/javax/mail/mail/1.4.4/mail-1.4.4.pom`.

I'm assuming this is a bug in Maven 3's core dependency resolution mechanisms (as opposed to maven-invoker-plugin) since Maven 2.2.1 doesn't exhibit the behaviour.

---

**Affects:** 3.0.2, 3.0.3

6 votes, 6 watchers

**Comment by jira-importer** [comment]
**[Chas Emerick](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cemerick)** commented

I mistakenly swapped the order of link and link text in my report, but it doesn't look like I have privs to edit the ticket  sorry!

**Comment by jira-importer** [comment]
**[David Matějček](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmatej)** commented

I have the same problem, deployed SNAPSHOT version from one machine to the repository server yesterday, but today another machine does not update the artifact, even with -U.
When I deleted maven-metadata-local.xml of the artifact, the dependency was updated (even without -U).
There wasn't any problems with Maven2.
Maven 3.0.3, Kubuntu 10.04 - 2.6.32-32-generic, Sun JDK 1.6.0.24.

**Comment by jira-importer** [comment]
**[David Matějček](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmatej)** commented

I have an idea - I switched between Maven2 and Maven3. Maybe if the artifact is downloaded by Maven2, Maven3 skips updating due to some incompatibility?

**Comment by jira-importer** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

The hotfix for this is to replace `<maven-home>/lib/aether-\*-1.11.jar`'s with their 1.12 equivalents.

**Comment by jira-importer** [comment]
**[Chas Emerick](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cemerick)** commented

Confirmed  using aether 1.12 with Maven 3.0.3 resolves this issue.

FYI for others hitting this, the only place I could find aether binaries was [central](http://repo1.maven.org/maven2/org/sonatype/aether/).  The artifacts in question are -api, -connector-wagon, -impl, -spi, and -util.

It looks like aether 1.12 was released on 5/30, and other notes around on the 'nets seem to indicate that aether 1.12 fixes a number of problems with maven 3.0.3.  Anything that can be done to get a v3.0.4 maven release out that includes aether 1.12 would be greatly appreciated.

Thanks :-)
